src/freenet/clients/fcp/FCPPluginClientTracker.java

```diff
@@ -0,0 +1,212 @@
+        public final UUID clientID;
+
+        public FCPPluginClientWeakReference(FCPPluginClient referent,
+            ReferenceQueue<FCPPluginClient> referenceQueue) {
+            
+            super(referent, referenceQueue);
+            clientID = referent.getID();
+        }
+   
+    }
+    
+    /**
+     * Stores the {@link FCPPluginClient} so in the future it can be obtained by its ID with
+     * {@link FCPPluginClientTracker#getClient(UUID)}.
+     * 
+     * <b>Must</b> be called for any newly created {@link FCPPluginClient} before passing it to
+     * {@link ServerSideFCPMessageHandler#handlePluginFCPMessage(FCPPluginClient,
+     * FCPPluginMessage)}.
+     * 
+     * Unregistration is not supported and not necessary.
+     */
+   void registerClient(FCPPluginClient client) {
+        clientsByIDLock.writeLock().lock();
+        try {
+            // No duplicate checks needed: FCPPluginClient.getID() is a random UUID.
+            clientsByID.put(client.getID(),
+                new FCPPluginClientWeakReference(client, disconnectedClientsQueue));
+        } finally {
+            clientsByIDLock.writeLock().unlock();
+        }
+    }
+    
+    /**
+     * For being used by implementors of {@link ServerSideFCPMessageHandler}.<br/>
+     * NOT for being used by clients: If you are a client using a {@link FCPPluginClient} to connect
+     * to a server plugin, you have to keep a reference to the {@link FCPPluginClient} in memory.
+     * <br/>
+     * This is necessary because this class only keeps {@link WeakReference}s to the
+     * {@link FCPPluginClient} objects. Once they are not referenced by a strong reference anymore,
+     * they will be garbage collected and thus considered as disconnected.<br/>
+     * The job of keeping the strong references is at the client.
+     * 
+     * @param clientID
+     *            The ID of{@link FCPPluginClient#getID()} of a client which has already sent a
+     *            message to your plugin via
+     *            {@link ServerSideFCPMessageHandler#handlePluginFCPMessage(FCPPluginClient,
+     *            FCPPluginMessage)}
+     * @return The client with the given ID, for as long as it is still connected to the node.
+     * @throws IOException
+     *             If there has been no client with the given ID or if it has disconnected already.
+     */
+    public FCPPluginClient getClient(UUID clientID) throws IOException {
+        FCPPluginClientWeakReference ref = null;
+        
+        clientsByIDLock.readLock().lock();
+        try {
+            ref = clientsByID.get(clientID);
+        } finally {
+            clientsByIDLock.readLock().unlock();
+        }
+        
+        FCPPluginClient client = ref != null ? ref.get() : null;
+        
+        if(client == null) {
+            throw new IOException("FCPPluginClient not found, maybe it has disconnected already."
+                + " ClientID: " + clientID);
```

Comment:
IOException is used wherever disconnection is the most likely explanation. This is done to follow the convention of Java Sockets, for example see: https://docs.oracle.com/javase/6/docs/api/java/net/Socket.html
The non-Socket replacement would be PluginNotFoundException as thats what PluginManager typically uses when trying to query a plugin, or one of its interfaces, such as notably the FCP interface which the pull request is about.

It was difficult to decide when to use PluginNotFoundException and when to use IOException.
I decided to transition to use IOException whenever disconnection CAN be the reason, and to use PluginNotFoundException when disconnection CANNOT be the reason.
Here disconnection can be the reason: The official way for a client plugin to close a FCPPluginClient connection is to discard all hard references to it. The sole purpose of this class is to detect that by monitoring the WeakReference tree. So it pretty much is about detecting disconnection, and thats why it throws IOException.
Consequently, the callers of this function are very simple, they all just throw it outward, and it makes more and more sense the higher you go in the hierarchy. In the end it will be thrown outside a send() function.

(An example of when disconnection is not relevant and thus PluginNotFoundException is thrown is PluginRespirator.connectToOtherPlugin())

ACK to keep as is?